### ISSUE TITLE
Preview window improvements

### ISSUE THREAD

**Issue by Flavelius** [KybernetikGames/animancer#100]
# Use Case
Previewing multiple layer interactions, looping transition previews, removing clutter

# Solution

Layers ca be created in the preview, but additional ones serve no purpose as clips/states/transitions cannot be assigned to them. Making that prossible would allow previewing interactions of multiple layers, like additive ones.

Currently it's also not possible to simply loop the transition preview, so it has to be restarted by hand every time it ends, having a little checkbox for that would help.
Also, deleting layers or destroying animation states via rightclick throws a bunch of errors as this is likely not intended here.

Additionally, removing the 'Previous Animation' field will probably remove some confusion, as the 'current animation' doesn't control how others fade into it (only how it does into others). Currently it plays 3 clips if those are set but viewing the first uncontrollable transition has little to no use (or i just don't see it).

It would also probably help for visual navigation to put some spacing between the clip/transition selections and the model dropdown (and maybe some other elements). Currently it's a little dense with alot of information (some unrelated) packed together

**Comment by Flavelius** [comment]
A similar part that could be improved is the transition property drawer. At runtime it tanks performance somehow. In my script i only have 1 linearmixer transition with 4 clips and 3 additional clip transition fields and fps drops a lot when an object with it is inspected.
lowering the redraw rate doesn't fix it, it just stutters at a lower frequency.

**Comment by KybernetikGames** [comment]
> allowing previewing interactions of multiple layers

What sort of controls should the layers have and how would you like to set them up?

> Currently it's also not possible to simply loop the transition preview

That would be simple enough. Should the loop occur right when the Current Animation finishes fading out or should it have a bit of a delay?

> Additionally, removing the 'Previous Animation' field will probably remove some confusion, as the 'current animation' doesn't control how others fade into it

The `Fade Duration` of the transition you're previewing controls how long it takes to go from the `Previous Animation` to the `Current Animation`.

And the `End Time` determines when it starts fading to the `Next Animation` (but you have no control over how long it takes since that's not part of the transition).

So both the previous and next are necessary.

> put some spacing between the clip/transition selections and the model dropdown

Sounds reasonable.

> transition property drawer. At runtime it tanks performance somehow

Have you used the profiler with Profile Editor and Deep Profile turned on to make sure that's actually the cause of the problem? I get quite a few reports about Inspector efficiency which turn out to be caused by other things.

**Comment by Flavelius** [comment]
1. It could have a similar one to the current interface, with an additional dropdown to select a target layer (and maybe a time-offset field. This in turn nested in a list that allows to create multiple of those configurations (for each layer one wants to play something on) with one (undeletable) default entry that targets the base layer, so it behaves like it currently does by default. The play button would then start/stop all those layer previews together and use their timeoffsets accordingly.

2. The way i understand transitions from reading the documentation is that one of those still is unnecessary. Looks like i just inverted my statement intention and it's the next animation then. Because end-time is not part of the transition and the next animation controls when the end time actually occurs. The docs confirm that the end time even has no effect on the animation, so previewing that is unnecessary. It also just adds to confusion as the preview is currently actually showing 2 transitions where it should only preview the one that's relevant (like the animator controller does).

3. For me either way will work. Having it looping makes it just a little easier to tweak the values while not having to hit play repeatedly, a delay or not won't make a difference. But maybe, if it's not too much work, the loop range could be configurable with a min-max slider for example, so when previewing longer animations unnecessary parts could be skipped.

Transition Drawer: yes, i profiled it with deep profiling enabled. i even posted it in the odin inspector (thirdparty asset) discord to get some confirmation whether the overhead comes from odin or the drawer and they confirmed that odin merely took 5ms in my case from 27ms with 3 ms overhead but the drawer took 19ms. My script has 1 LinearMixerTransition with 4 clips and 4 additional single clip transition fields.

**Comment by KybernetikGames** [comment]
> The docs confirm that the end time even has no effect on the animation, so previewing that is unnecessary.

It doesn't automatically do anything, but starting a transition into something else is the most common thing its used for and there isn't really any better way to visualise it than to actually do a transition. I agree that it's a little confusing, but it's an important part of the transition data.

The only way I can really think of to improve it would require something like what I described in https://github.com/KybernetikGames/animancer/issues/80 so that it could use the end time of one transition and the other details of another so that it only needs previous and next.

> Transition Drawer

I can't see anything in it that looks particularly slow and it's not having any noticeable effect on my frame rate even if I make it repaint constantly.

Can you use Deep Profiling to narrow down which method is causing problems?

**Comment by Flavelius** [comment]
> It doesn't automatically do anything, but starting a transition into something else is the most common thing its used for and there isn't really any better way to visualise it than to actually do a transition. I agree that it's a little confusing, but it's an important part of the transition data.

But it's just a generic event which could also just play the next one without fading or in reverse or fade out the layer or something else and it doesn't preview that either or lets you configure it, so using this hidden default transition is somewhat arbitrary and only really helpful for the single usecase of doing a default transition, which is actually not that common because that's the reason transitions need a preview in the first place - to configure the fade time etc.

**Comment by Flavelius** [comment]
![grafik](https://user-images.githubusercontent.com/8841352/108484748-7054c380-729c-11eb-973f-1b972df568a6.png)
Here you can see the difference (at the top) for when the drawer is drawn and before it was focused

**Comment by KybernetikGames** [comment]
That `Drawer.OnGUI` method is only taking about 7ms for me even with about 10 states in the list and most of that is caused by the calls to Unity's GUI methods like `EditorGUI.ObjectField`.

Can you see anything unexpectedly costly in the methods under that?

Also, can you show me a screenshot of your Inspector so I can set it up the same?

**Comment by Flavelius** [comment]
![grafik](https://user-images.githubusercontent.com/8841352/108587726-8f635c00-7355-11eb-967b-5aec806b92ec.png)
This is the part focused on the OnGUI call (which is like 40 calls deep).
My guess is that it may be related to the amount of garbage allocated per frame (75kb total) which the drawer contributes a good part to. Maybe this could be eliminated with an UIElements based drawer.

**Comment by Flavelius** [comment]
oh and there's another animancer call:
![grafik](https://user-images.githubusercontent.com/8841352/108588023-4f9d7400-7357-11eb-9e35-05773ec8b393.png)

**Comment by Flavelius** [comment]
A large part of the performance drop seems to be just unoptimized unity stuff too though, animancer's part is ~25ms from ~90ms. With the components collapsed (so they don't draw) the difference is still 22ms without any object focused (inspector clear) to ~60ms with the object focused and collapsed components.
It still adds a noticable drop, as seen here:

https://user-images.githubusercontent.com/8841352/108588548-1b778280-735a-11eb-8214-7b0d0a537b9b.mp4

**Comment by KybernetikGames** [comment]
Yeah, garbage could well be the problem but I'm already using stuff like my `ConversionCache` class to avoid it as much as possible in my own methods so I haven't found any room for significant improvement.

UI Elements might help, but I expect it would take quite a lot of work to convert all my existing drawers over (which means work I'm not putting into other parts of the system). In particular, setting up something procedurally drawn like the Animancer Event timeline would likely be a massive pain and I don't know how well the system can handle inheritance (since every transition type has its own drawer type).
